shotover-proxy/src/transforms/cassandra/sink_cluster.rs

```diff
@@ -62,92 +84,244 @@ impl CassandraSinkCluster {
-                            ..
-                        } => {
-                            if let Some(raw_bytes) = message.as_raw_bytes() {
-                                if let Ok(Opcode::Error) =
-                                    cassandra::raw_frame::get_opcode(raw_bytes)
-                                {
-                                    self.failed_requests.increment(1);
-                                }
-                            }
-                            responses.push(message);
-                        }
-                        Response {
-                            mut original,
-                            response: Err(err),
-                        } => {
-                            original.set_error(err.to_string());
-                            responses.push(original);
+        // Attempt to populate nodes list if we still dont have one yet
+        if self.nodes.is_empty() {
+            let nodes_shared = self.nodes_shared.read().await;
+            self.nodes.extend(nodes_shared.iter().cloned());
+        }
+
+        if !self.init_handshake_complete {
+            for message in &messages {
+                if let Some(last) = self.init_handshake.last_mut() {
+                    if let Some(Frame::Cassandra(CassandraFrame {
+                        operation: CassandraOperation::AuthResponse(_),
+                        ..
+                    })) = last.frame()
+                    {
+                        // Only send a handshake if the task really needs it
+                        // i.e. when the channel of size 1 is empty
+                        if let Ok(permit) = self.task_handshake_tx.try_reserve() {
+                            permit.send(TaskHandshake {
+                                handshake: self.init_handshake.clone(),
+                                address: self.init_handshake_address.as_ref().unwrap().clone(),
+                            })
                         }
-                    };
-                }
-                Ok(None) => break,
-                Err(_) => {
-                    error!(
-                        "timed out waiting for responses, received {:?} responses but expected {:?} responses",
-                        responses.len(),
-                        expected_size
-                    );
+                        self.init_handshake_complete = true;
+                        break;
+                    }
                 }
+                self.init_handshake.push(message.clone());
             }
         }
 
+        let mut responses_future = FuturesOrdered::new();
+        for message in messages {
+            let (return_chan_tx, return_chan_rx) = oneshot::channel();
+            if self.nodes.is_empty() || !self.init_handshake_complete {
+                self.contact_point_connection.as_mut().unwrap()
+            } else {
+                let random_node = self.nodes.choose_mut(&mut self.rng).unwrap();
```

Comment:
What are your plans for abstracting the query planning/routing strategies?

Reply:
I didnt know there were any current plans to allow multiple planning/routing strategies?

Reply:
Oh true, I just assumed we would. e.g.

https://docs.rs/cdrs-tokio/latest/cdrs_tokio/load_balancing/index.html